### PR TITLE
Skip Iceberg row filter when not needed

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

Avoid recreating Iceberg Pages for the DeleteFilter when not needed.

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement/Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Iceberg Connector

> How would you describe this change to a non-technical end user or system administrator?

Small performance improvement

## Related issues, pull requests, and links

https://github.com/trinodb/trino/pull/11642

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 1]
```diff
@@ -256,7 +256,7 @@ public ConnectorPageSource createPageSource(
                 requiredColumns,
                 dataPageSource.get(),
                 projectionsAdapter,
-                deleteFilter);
+                deleteFilter.hasPosDeletes() || deleteFilter.hasEqDeletes() ? Optional.of(deleteFilter) : Optional.empty());
```
Comment (electrum) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:259]: Can shorten to
```java
Optional.of(deleteFilter).filter(filter -> filter.hasPosDeletes() || filter.hasEqDeletes());
```
Or change above to
```java
Optional<DeleteFilter<TrinoRow>> deleteFilter = Optional.of(
        new TrinoDeleteFilter(dummyFileScanTask, tableSchema, requiredColumns, fileIO))
        .filter(filter -> filter.hasPosDeletes() || filter.hasEqDeletes());

**Review by electrum — APPROVED** [review]

**Comment by alexjo2144** [comment]
Updated to use Optional.filter. Thanks for reviewing

**Review by findinpath — APPROVED** [review]

**Comment by alexjo2144** [comment]
@electrum can we merge this when you get a chance?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 2]
```diff
@@ -256,7 +256,7 @@ public ConnectorPageSource createPageSource(
                 requiredColumns,
                 dataPageSource.get(),
                 projectionsAdapter,
-                deleteFilter);
+                Optional.of(deleteFilter).filter(filter -> filter.hasPosDeletes() || filter.hasEqDeletes()));
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:259]: Why is it needed?

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:259]: The data Page is always rewritten in the IcebergPageSource right now, this makes it conditional on if there are actually any deletes to account for.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:259]: sorry for not being clear. I understand what the change is for.
wonder why `Optional#filter` is needed here.
if "deleteFilter" is not needed, why was it passed until here? did the split source attach some useless information to the split?

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:259]: Oh, I see. We could also make this conditional on `split.getDeletes().isEmpty()`  if that's clearer? They're equivalent

Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:259]: > why was it passed until here?

The DeleteFilter is constructed on the line right above this, the only thing passed around until here are some empty lists. `split.getDeletes()` should be empty, as should `deleteFilterRequiredSchema`

**Pull request merged**